File: OneQuestionPerDay/LCOF59/max_sliding_window.py
from typing import List
from collections import deque


class Solution:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        dq = deque()
        res = []
        for i, n in enumerate(nums):
            if len(dq) > 0 and dq[0] <= i - k:
                dq.popleft()
            while dq and nums[dq[-1]] < n:
                dq.pop()

            dq.append(i)
            if i >= k - 1:
                res.append(nums[dq[0]])
        return res

    def test(self):

        for nums, k, target in [
            ([7, 2, 4], 2, [7, 4])
        ]:
            ans = self.maxSlidingWindow(nums, k)
            assert ans == target, f"target: {target}, ans: {ans}"


if __name__ == "__main__":
    Solution().test()

File: OneQuestionPerDay/LCOF59/max_sliding_window.go
package main

func maxSlidingWindow(nums []int, k int) []int {
	var window, res []int
	for i := 0; i < len(nums); i++ {
		if len(window) > 0 && window[0] <= i-k {
			window = window[1:]
		}
		for len(window) > 0 && nums[window[len(window)-1]] < nums[i] {
			window = window[:len(window)-1]
		}
		window = append(window, i)
		if i >= k-1 {
			res = append(res, nums[window[0]])
		}
	}
	return res
}
